### ISSUE TITLE
Can I use joern for Java?

### ISSUE THREAD

**Issue by nashid** [joernio/joern#540]
I want to use joern for java program analysis. However, it appears joern supports only C/C++.

Is this understanding correct?

Or joern can be used for java as well?

**Comment by DavidBakerEffendi** [comment]
Joern currently uses [Plume](https://plume-oss.github.io/plume-docs/) to support JVM byte code. Note that this only analyses the byte code and doesn't use the source code - so the results are linked back to source code via line number and variable name information (if compiled with debugging info on) and/or method/class names.

It is another open-source project started on it's own but has recently been making closer and closer migrations to becoming more part of Joern.io projects.

**Comment by nashid** [comment]
@DavidBakerEffendi so [Plume](https://plume-oss.github.io/plume-docs/plume-basics/extracting-cpg/) does not provide API access to Code Property Graph for a java program?

My intention has been to use the Code Property Graph and using that to perform def-use analysis.

**Comment by DavidBakerEffendi** [comment]
Yeah so you can give Java class files or a JAR to Joern (which should then switch to Plume) and one can do the usual analysis on the projected CPG - the only limitation is that the AST will look much more different than the source code but the flow should be the same.

**Comment by nashid** [comment]
@DavidBakerEffendi in this case how I can correlate the projected CPG back to the source code? Is there any way I can relate it back to the source code? 

If I cant go back to the source code, it would not help me 😭 

Basically, I want to extract all the def-use links from the CPG and relate those with the variables. 

Any pointers, please.

**Comment by DavidBakerEffendi** [comment]
So if you give Plume source code, it will compile it with debugging information on - the only issue in this case is that the code must be compilable. If, for instance, you had the following code:

```java
13. public boolean authenticate(String username, String password) {
14.    var query = "SELECT * FROM accounts WHERE username = " + username + " AND password = " + password;
15.    stmt = conn.createStatement();
16.    ResultSet rs = stmt.executeQuery(query);
17.    return rs.hasNext();
18. }
```

With debugging information you would be able to see that there is a data-flow from potentially tainted `param1` and `param2` of method `authenticate` at line 13 (fully qualified path available by the `methodFullName` property) to a sensitive sink `java.sql.Statement.executeQuery` at line 16.

Without debugging information you'll have the fully qualified method information `authenticate` with line number offset of some call with callee `java.sql.Statement.executeQuery` within the method.

**Comment by DavidBakerEffendi** [comment]
The advantage of Plume is that you can hack away at close-source programs 😉 but there is a lot of use for fuzzy-parsers when you're trying to develop the code. These are on their way though

**Comment by nashid** [comment]
> So if you give Plume source code, it will compile it with debugging information on - the only issue in this case is that the code must be compilable. If, for instance, you had the following code:
> 
> ```java
> 13. public boolean authenticate(String username, String password) {
> 14.    var query = "SELECT * FROM accounts WHERE username = " + username + " AND password = " + password;
> 15.    stmt = conn.createStatement();
> 16.    ResultSet rs = stmt.executeQuery(query);
> 17.    return rs.hasNext();
> 18. }
> ```
> 
> With debugging information you would be able to see that there is a data-flow from potentially tainted `param1` and `param2` of method `authenticate` at line 13 (fully qualified path available by the `methodFullName` property) to a sensitive sink `java.sql.Statement.executeQuery` at line 16.
> 
> Without debugging information you'll have the fully qualified method information `authenticate` with line number offset of some call with callee `java.sql.Statement.executeQuery` within the method.

I have no problem running Plume with debugging enabled as I would run this step in a pre-processing phase. However, I am wondering whether I would get API access from my Java code? 

Do you have any examples how to achieve this in Plume?

I am using [ManySStuBs4J](https://github.com/mast-group/mineSStuBs) dataset. So I have access to the actual source code and I want to incorporate def-use relation as an example.

**Comment by DavidBakerEffendi** [comment]
Ah I understand, so the Plume organization repository has examples of running on Plume directly using Java for its various supported graph database backends. This can be found at www.github.com/plume-oss/plume-examples.

**Comment by DavidBakerEffendi** [comment]
In addition to this there is documentation to how these phases work - the load method accepts JARs and directories too - you can call `load` multiple times to load the whole application and then call `project` which will the consume all the loaded files.

This will probably make more sense once you've seen the examples on the repo.

**Comment by DavidBakerEffendi** [comment]
A very cool dataset by the way - are you looking at training a GNN to learn program vulnerabilities? It's something I've proposed in my PhD which is a part of why Plume is developed.

**Comment by nashid** [comment]
> www.github.com/plume-oss/plume-examples

I was actually looking into the examples. I am wondering is there a way just to run the program analysis part i.e. just extracting the CPG without any graph database backend?

**Comment by nashid** [comment]
> A very cool dataset by the way - are you looking at training a GNN to learn program vulnerabilities? It's something I've proposed in my PhD which is a part of why Plume is developed.

yeah - that's the intention. Super awesome to hear that about your plan! It must be super exciting.

**Comment by DavidBakerEffendi** [comment]
> > [www.github.com/plume-oss/plume-examples](http://www.github.com/plume-oss/plume-examples)
> 
> I was actually looking into the examples. I am wondering is there a way just to run the program analysis part i.e. just extracting the CPG without any graph database backend?

The best would be the OverflowDb driver since that is what Joern does with C/C++ anyway. It's an in-memory graph database with a swap-to-disk mechanism if memory is limited. The `.bin` is very compact and has a really nice compression ratio if you use something like LZMA.

**Comment by DavidBakerEffendi** [comment]
> > A very cool dataset by the way - are you looking at training a GNN to learn program vulnerabilities? It's something I've proposed in my PhD which is a part of why Plume is developed.
> 
> yeah - that's the intention. Super awesome to hear that about your plan! It must be super exciting.

Yeah really looking forward to it - btw just got news that the framework for the Java fuzzy parser has been finalised so that will now start on Joern.io

**Comment by DavidBakerEffendi** [comment]
When it is created then we'll make a post - stay tuned!

**Comment by nashid** [comment]
@DavidBakerEffendi any idea about the tentative timeline, please?
